go/test/endtoend/recovery/shardedpitr/shardedpitr_test.go

```diff
@@ -0,0 +1,538 @@
+	// +----+--------+------------+
+	recoveryTablet(t, bs1, "2m", restoreKS3Name, "80-", "INT64(7)")
+
+}
+
+func performResharding(t *testing.T) {
+	err := clusterInstance.VtctlclientProcess.ApplyVSchema(keyspaceName, vSchema)
+	require.NoError(t, err)
+
+	err = clusterInstance.VtctlProcess.ExecuteCommand("InitShardMaster", "-force", "ks/-80", shard0Master.Alias)
+	require.NoError(t, err)
+
+	err = clusterInstance.VtctlProcess.ExecuteCommand("InitShardMaster", "-force", "ks/80-", shard1Master.Alias)
+	require.NoError(t, err)
+
+	// we need to create the schema, and the worker will do data copying
+	for _, keyspaceShard := range []string{"ks/-80", "ks/80-"} {
+		err = clusterInstance.VtctlclientProcess.ExecuteCommand("CopySchemaShard", "ks/0", keyspaceShard)
+		require.NoError(t, err)
+	}
+
+	err = clusterInstance.VtctlclientProcess.ExecuteCommand("SplitClone", "ks", "0", "-80,80-")
+	require.NoError(t, err)
+
+	err = clusterInstance.VtctlclientProcess.ExecuteCommand("MigrateServedTypes", "ks/0", "rdonly")
+	require.NoError(t, err)
+
+	err = clusterInstance.VtctlclientProcess.ExecuteCommand("MigrateServedTypes", "ks/0", "replica")
+	require.NoError(t, err)
+
+	// then serve master from the split shards
+	err = clusterInstance.VtctlclientProcess.ExecuteCommand("MigrateServedTypes", "ks/0", "master")
+	require.NoError(t, err)
+
+	// remove the original tablets in the original shard
+	removeTablets(t, []*cluster.Vttablet{master, replica})
+
+	for _, tablet := range []*cluster.Vttablet{replica} {
+		err = clusterInstance.VtctlclientProcess.ExecuteCommand("DeleteTablet", tablet.Alias)
+		require.NoError(t, err)
+	}
+	err = clusterInstance.VtctlclientProcess.ExecuteCommand("DeleteTablet", "-allow_master", master.Alias)
+	require.NoError(t, err)
+
+	// rebuild the serving graph, all mentions of the old shards should be gone
+	err = clusterInstance.VtctlclientProcess.ExecuteCommand("RebuildKeyspaceGraph", "ks")
+	require.NoError(t, err)
+
+	// delete the original shard
+	err = clusterInstance.VtctlclientProcess.ExecuteCommand("DeleteShard", "ks/0")
+	require.NoError(t, err)
+
+	// Restart vtgate process
+	err = clusterInstance.VtgateProcess.TearDown()
+	require.NoError(t, err)
+
+	err = clusterInstance.VtgateProcess.Setup()
+	require.NoError(t, err)
+
+	clusterInstance.WaitForTabletsToHealthyInVtgate()
+}
+
+func startBinlogServer(t *testing.T, masterTablet *cluster.Vttablet) *binLogServer {
+	bs, err := newBinlogServer(hostname, clusterInstance.GetAndReservePort())
+	require.NoError(t, err)
+
+	err = bs.start(mysqlMaster{
+		hostname: binlogHost,
+		port:     masterTablet.MysqlctlProcess.MySQLPort,
+		username: mysqlUserName,
+	})
+	require.NoError(t, err)
+	return bs
+}
+
+func removeTablets(t *testing.T, tablets []*cluster.Vttablet) {
+	var mysqlProcs []*exec.Cmd
+	for _, tablet := range tablets {
+		proc, _ := tablet.MysqlctlProcess.StopProcess()
+		mysqlProcs = append(mysqlProcs, proc)
+		tablet.VttabletProcess.TearDown()
```

Comment:
This might throw error, we should stop tablet process after the mysqls are down.

Reply:
Move the teardown func after all the mysqls are down